src/tools/miri/src/machine.rs

```diff
@@ -967,11 +972,26 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
             )
         });
         let buffer_alloc = ecx.machine.weak_memory.then(weak_memory::AllocState::new_allocation);
-        let alloc: Allocation<Provenance, Self::AllocExtra> = alloc.adjust_from_tcx(
+
+        let mut alloc: Allocation<Provenance, Self::AllocExtra> = alloc.adjust_from_tcx(
             &ecx.tcx,
-            AllocExtra { borrow_tracker, data_race: race_alloc, weak_memory: buffer_alloc },
+            AllocExtra {
+                borrow_tracker,
+                data_race: race_alloc,
+                weak_memory: buffer_alloc,
+                backtrace: None,
+            },
             |ptr| ecx.global_base_pointer(ptr),
         )?;
+        // If an allocation is leaked, we want to report a backtrace to indicate where it was
+        // allocated. Stack and CallerLocation can't be leaked, so we can skip recording a
+        // backtrace for those memory kinds.
+        if let MemoryKind::Machine(_) = kind {
```

Comment:
This made me think it would be neat to encode in the `MemoryKind`, but that seems annoying to handle.